File: sdk/dart-holons/lib/src/relay.dart
import 'dart:async';
import 'dart:io' as io;

import 'package:fixnum/fixnum.dart';
import 'package:grpc/grpc.dart';
import 'package:grpc/service_api.dart' as grpc_api;

import 'gen/relay/v1/relay.pbgrpc.dart';
import 'observability.dart' as observability;

class RelayOptions {
  const RelayOptions({this.downstreamConn});

  final grpc_api.ClientChannel? downstreamConn;
}

Service relayService(RelayOptions options) => _RelayServer(options);

class _RelayServer extends RelayServiceBase {
  _RelayServer(this.options);

  final RelayOptions options;
  var _received = 0;

  @override
  Future<TickResponse> tick(ServiceCall call, TickRequest request) async {
    _received += 1;
    final obs = observability.current();
    final slug = _responderSlug(obs);
    final uid = obs.cfg.instanceUid;
    obs.logger('tick').info('tick received', fields: {
      'sender': request.sender,
      'note': request.note,
      'responder_slug': slug,
      'responder_uid': uid,
    });
    obs.counter(
      'cascade_ticks_total',
      help: 'Ticks received by this cascade node.',
      labels: {'responder_uid': uid},
    )?.inc();

    final hops = <HopReceipt>[];
    final downstream = options.downstreamConn;
    if (downstream != null) {
      final response = await RelayServiceClient(downstream).tick(request);
      hops.addAll(response.hops);
    }
    hops.add(HopReceipt(
      slug: slug,
      uid: uid,
      received: Int64(_received),
    ));
    return TickResponse(
      responderSlug: slug,
      responderInstanceUid: uid,
      hops: hops,
    );
  }
}

String _responderSlug(observability.Observability obs) {
  final configured = obs.cfg.slug.trim();
  if (configured.isNotEmpty) {
    return configured;
  }
  return io.Platform.resolvedExecutable.split(io.Platform.pathSeparator).last;
}

File: sdk/rust-holons/src/relay.rs
use crate::gen::relay::v1 as pb;
use crate::observability::{self, Field};
use std::collections::BTreeMap;
use std::sync::atomic::{AtomicI64, Ordering};
use tonic::{Request, Response, Status};

#[derive(Default)]
pub struct RelayServer {
    downstream: Option<tonic::transport::Channel>,
    received: AtomicI64,
}

impl RelayServer {
    pub fn new(downstream: Option<tonic::transport::Channel>) -> Self {
        Self {
            downstream,
            received: AtomicI64::new(0),
        }
    }
}

pub fn service(
    downstream: Option<tonic::transport::Channel>,
) -> pb::relay_service_server::RelayServiceServer<RelayServer> {
    pb::relay_service_server::RelayServiceServer::new(RelayServer::new(downstream))
}

#[tonic::async_trait]
impl pb::relay_service_server::RelayService for RelayServer {
    async fn tick(
        &self,
        request: Request<pb::TickRequest>,
    ) -> Result<Response<pb::TickResponse>, Status> {
        let count = self.received.fetch_add(1, Ordering::SeqCst) + 1;
        let request = request.into_inner();
        let obs = observability::current();
        let slug = responder_slug(&obs);
        let uid = obs.cfg.instance_uid.clone();
        obs.logger("tick").info(
            "tick received",
            &[
                ("sender", Field::String(request.sender.clone())),
                ("note", Field::String(request.note.clone())),
                ("responder_slug", Field::String(slug.clone())),
                ("responder_uid", Field::String(uid.clone())),
            ],
        );
        let mut labels = BTreeMap::new();
        labels.insert("responder_uid".to_string(), uid.clone());
        if let Some(counter) = obs.counter(
            "cascade_ticks_total",
            "Ticks received by this cascade node.",
            labels,
        ) {
            counter.inc();
        }

        let mut hops = Vec::new();
        if let Some(channel) = self.downstream.clone() {
            let mut client = pb::relay_service_client::RelayServiceClient::new(channel);
            let response = client.tick(request).await?.into_inner();
            hops.extend(response.hops);
        }
        hops.push(pb::HopReceipt {
            slug: slug.clone(),
            uid: uid.clone(),
            received: count,
        });

        Ok(Response::new(pb::TickResponse {
            responder_slug: slug,
            responder_instance_uid: uid,
            hops,
        }))
    }
}

fn responder_slug(obs: &observability::Observability) -> String {
    let configured = obs.cfg.slug.trim();
    if !configured.is_empty() {
        return configured.to_string();
    }
    "unknown".to_string()
}
